---
title: "Testis_bulk Deconv"
author: "BH"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r packages}

if (!require("pacman")) install.packages("pacman")
pacman::p_load(dplyr, readxl, ggplot2, ggsci, ggthemes, vroom, readr, tidyverse, DWLS, stringr, Metrics, quadprog)

theme_set(theme_few() + theme(text = element_text(size=28, family="Myriad Pro")))

```



```{r load PBs}

testisrna <- readRDS("~/OneDrive - UW/data for deconv/all_testis_rnaseq.rds")

metadata <- read_excel("~/OneDrive - UW/data for deconv/Paper revision materials/DATA/cpm/Testis/testis_metadata.xlsx")

germ_cols <- c("d_spg", "SSCs", "Germ", "ProSGSSC", "Pro")
leydig_cols <- c("Leydig", "StromaLeydig")



unlist_func <- function(list) {bind_rows(lapply(names(list), function(name) {
    df <- list[[name]]
    df$source <- name
    existing_germ_cols <- germ_cols[germ_cols %in% names(df)]
    existing_lc_cols <- leydig_cols[leydig_cols %in% names(df)]


  refs <- unlist(strsplit(name, "_"))  
  df %>%
    mutate(#pb = sapply(strsplit(refs, "_"), `[`, 1),
           #pbref = sapply(strsplit(refs, "_"), `[`, 2),
          source = name,
           imp = refs[1],
           sigref = refs[2],
          germ_all = rowSums(across(all_of(existing_germ_cols)), na.rm = TRUE),
          lc_all = rowSums(across(all_of(existing_lc_cols)), na.rm = TRUE)) %>%
  select(-all_of(existing_germ_cols)) %>% select(-all_of(existing_lc_cols))
  }))}




```


Load the signatures matrices
```{r sigs}
#important to use the corrected folder

swd <- list.dirs("~/OneDrive - UW/data for deconv/Signature matrices impute+no impute (D)/Testis/DWLS_testis/alldata_revision_04132025/", full.names = TRUE)

# Initialize an empty list to hold sublists for each type
all_matrices <- list()

type <- c("noIMP", "ALRA")

for (dir in swd) {

    for (t in type) {
        if (grepl(t, dir)) {
        
            if (is.null(all_matrices[[t]])) {
                all_matrices[[t]] <- list() }
            
            rds_files <- list.files(dir, pattern = "Sig.rds", full.names = TRUE, recursive = TRUE)
            
            for (rds_file in rds_files) {
            
                data <- readRDS(rds_file)
                file_name_part <- basename(dirname(rds_file))
                new_name <- paste0(t, "_", file_name_part, "_sig")
                all_matrices[[t]][[new_name]] <- data   }} }}

```

```{r OLS}

OLS <- data.frame()

runOLSall <- function(sig_matrix_in, df_in) {
  df <- df_in 
  rows <- intersect(rownames(sig_matrix_in), rownames(df))
  df <- df[rows, ]
  sig_matrix_in <- sig_matrix_in[rows, ]
  sig_matrix_in <- sig_matrix_in[, !(colnames(sig_matrix_in) %in% c("Endo", "Immune", "meiosis_spcs"))]

  samples <- colnames(df)
  results_matrix <- matrix(NA, nrow = length(samples), ncol = ncol(sig_matrix_in))
  rownames(results_matrix) <- samples
  colnames(results_matrix) <- colnames(sig_matrix_in)

  
  solveOLS <- function(S, B) {
    D <- t(S) %*% S
    d <- t(S) %*% B
    A <- cbind(diag(dim(S)[2]))
    bzero <- rep(-.05, dim(S)[2])
    solution<-solve.QP(D,d,A,bzero)$solution
    names(solution)<-colnames(S)
    print(round(solution/sum(solution),5))
   return(solution/sum(solution))
  }

  for (i in seq_along(samples)) {
    sample_name <- samples[i]
    B <- as.numeric(df[, sample_name])
    S <- as.matrix(sig_matrix_in)

    # actual deconv
    result <- solveOLS(S, B)
    results_matrix[sample_name, ] <- result
  }
  
  results_df <- as.data.frame(results_matrix)
  return(results_df)
}



alra_OLS <- unlist_func(lapply(all_matrices$ALRA, runOLSall, df_in = testisrna))


```




```{r DWLS}

runDWLSall <- function(df_in, sig_matrix_in) {
  # Convert gene column to rownames
  df <- df_in 

  # Intersect genes
  rows <- intersect(rownames(sig_matrix_in), rownames(df))
  df <- df[rows, ]
  sig_matrix <- sig_matrix_in[rows, ]
  sig_matrix <- sig_matrix[, !(colnames(sig_matrix) %in% c("Endo", "Immune", "meiosis_spcs"))]

  samples <- colnames(df)
  results_matrix <- matrix(NA, nrow = length(samples), ncol = ncol(sig_matrix))
  rownames(results_matrix) <- samples
  colnames(results_matrix) <- colnames(sig_matrix)

  solveOLSInternal<-function(S,B){
  D<-t(S)%*%S
  d<-t(S)%*%B
  A<-cbind(diag(dim(S)[2]))
  bzero<-c(rep(-0.05,dim(S)[2]))
  solution<-solve.QP(D,d,A,bzero)$solution
  names(solution)<-colnames(S)
  return(solution)
}
  #this is from DWLS github
solveDampenedWLS<-function(S,B){
  #first solve OLS, use this solution to find a starting point for the weights
  solution<-solveOLSInternal(S,B)
  #now use dampened WLS, iterate weights until convergence
  iterations<-0
  changes<-c()
  #find dampening constant for weights using cross-validation
  j<-findDampeningConstant(S,B,solution)
  change<-1
  while(change>.01 & iterations<1000){
    newsolution<-solveDampenedWLSj(S,B,solution,j)
    #decrease step size for convergence
    solutionAverage<-rowMeans(cbind(newsolution,matrix(solution,nrow = length(solution),ncol = 4)))
    change<-norm(as.matrix(solutionAverage-solution))
    solution<-solutionAverage
    iterations<-iterations+1
    changes<-c(changes,change)
  }
  print(round(solution/sum(solution),5))
  return(solution/sum(solution))
}


  for (i in seq_along(samples)) {
    sample_name <- samples[i]
    B <- as.numeric(df[, sample_name])
    S <- as.matrix(sig_matrix)

#Actual deconv
    result <- solveDampenedWLS(S, B)

    # Store the result in the appropriate place
    results_matrix[sample_name, ] <- result
  }

  # Convert the matrix to a data frame
  results_df <- as.data.frame(results_matrix)
  return(results_df)
}



alra_dwls <- unlist_func(lapply(all_matrices$ALRA, runDWLSall, df_in = testisrna))
noIMP_dwls <- unlist_func(lapply(all_matrices$noIMP, runDWLSall, df_in = testisrna))



dwls_testis <- rbind(alra_dwls,noIMP_dwls)%>% mutate(method='dwls')%>% rownames_to_column('ref')%>%
  mutate(ref = sub("\\..*$", "", ref)) %>%
  left_join(metadata, by = 'ref') %>%
  rename_with(~ c("Germ cells", "Leydig/Stroma cells", "Peritubular cells (PTM)", "Sertoli cells"), 
              .cols = c("germ_all", "lc_all", "PTM", "Sertoli")) %>%
  pivot_longer(
    cols = c("Peritubular cells (PTM)", "Sertoli cells", "Germ cells", "Leydig/Stroma cells"),
    values_to = 'expression' ) %>%
  mutate(expression = ifelse(expression < 0, 0, expression),
    shape = ifelse(expression == 0, "X", "point"),
    sigref = case_when(
      str_detect(source, "112") ~ "h112",
      str_detect(source, "120") ~ "h120",
      str_detect(source, "124") ~ "h124",
      str_detect(source, "d2") ~ "md2",
      str_detect(source, "d7") ~ "md7",
      TRUE ~ NA_character_
    )
  )


dwls_testis$lhtime <- factor(dwls_testis$lhtime, levels = c('0' ,'3' ,'5' ,'7' ,'10', '12', "na"))
dwls_testis$name <- factor(dwls_testis$name)

```




```{r dwls alra}

dwls_timeline <- ggplot(data=dwls_testis%>% filter( imp== 'ALRA' & lhtime != "na"), aes(x=lhtime, y=expression)) +
  
  geom_point(aes(fill=sigref,color = sigref, shape = shape), size= 6)+
  scale_fill_d3( breaks=c('h112', 'h120' ,'h124' , 'md2' , 'md7'), labels=c("Adult-H", "Infant-120", "Infant-124", "Day 2-M","Day 7-M")) + scale_color_d3( breaks=c('h112', 'h120' ,'h124' , 'md2' , 'md7'), labels=c("Adult-H", "Infant-120", "Infant-124", "Day 2-M","Day 7-M"))+
  geom_boxplot(size=1.2, alpha = 0.35, fill='lightgrey') + 
  theme_classic(base_size = 30, base_family = "Myriad Pro") +
  theme(legend.position="bottom", axis.text.x = element_text(angle = 45, hjust = 1)) +
  xlab("Day in vitro")+
  ylab("Estimated Cell Proportion")+
  scale_y_continuous(breaks = c(0,0.25, 0.5, 0.75, 1), limits = c(-0.05,1))+
  scale_shape_manual(values = c("X" = 4, "point" = 23))+
  labs(color = "Single Cell Reference") +
  guides(color = guide_legend(title.position = "bottom"))+
  facet_wrap(~name, nrow=2, ncol = 2)+
  theme(panel.spacing = unit(1.2, "lines"))
dwls_timeline

#ggsave(dwls_timeline, dpi = 600, filename = "~/OneDrive - UW/data for deconv/Paper revision materials/updated_figures/testis_timeline_dwls_ALRA.png", width = 12, height=14, units = 'in', bg = 'white')



```


```{r dwls noIMP}

dwls_noimp_timeline <- ggplot(data=dwls_testis%>% filter(imp== 'noIMP' & lhtime != "na"), aes(x=lhtime, y=expression)) +
  
  geom_point(aes(fill=sigref,color = sigref, shape = shape), size= 6)+
  scale_fill_d3( breaks=c('h112', 'h120' ,'h124' , 'md2' , 'md7'), labels=c("Adult-H", "Infant-120", "Infant-124", "Day 2-M","Day 7-M")) + scale_color_d3( breaks=c('h112', 'h120' ,'h124' , 'md2' , 'md7'), labels=c("Adult-H", "Infant-120", "Infant-124", "Day 2-M","Day 7-M"))+
  geom_boxplot(size=1.2, alpha = 0.35, fill='lightgrey') + 
  theme_classic(base_size = 30, base_family = "Myriad Pro") +
  theme(legend.position="bottom", axis.text.x = element_text(angle = 45, hjust = 1)) +
  xlab("Day in vitro")+
  ylab("Estimated Cell Proportion")+
  scale_y_continuous(breaks = c(0,0.25, 0.5, 0.75, 1), limits = c(-0.05,1))+
  scale_shape_manual(values = c("X" = 4, "point" = 23))+
  labs(color = "Single Cell Reference") +
  guides(color = guide_legend(title.position = "bottom"))+
  facet_wrap(~name, nrow=2, ncol = 2)+
  theme(panel.spacing = unit(1.2, "lines"))
dwls_timeline

#ggsave(dwls_noimp_timeline, dpi = 600, filename = "~/OneDrive - UW/data for deconv/Paper revision materials/updated_figures/testis_timeline_dwls_noIMP.png", width = 12, height=14, units = 'in', bg = 'white')


```


```{r OLS alra}

alra_OLS

ols_plot <- alra_OLS %>%
  rownames_to_column('ref') %>%
  mutate(ref = sub("\\..*$", "", ref)) %>%
  left_join(metadata, by = 'ref') %>%
  rename_with(~ c("Germ cells", "Leydig/Stroma cells", "Peritubular cells (PTM)", "Sertoli cells"), 
              .cols = c("germ_all", "lc_all", "PTM", "Sertoli")) %>%
  pivot_longer(
    cols = c("Peritubular cells (PTM)", "Sertoli cells", "Germ cells", "Leydig/Stroma cells"),
    values_to = 'expression' ) %>%
  mutate(expression = ifelse(expression < 0, 0, expression),
    shape = ifelse(expression == 0, "X", "point"),
    sigref = case_when(
      str_detect(source, "112") ~ "h112",
      str_detect(source, "120") ~ "h120",
      str_detect(source, "124") ~ "h124",
      str_detect(source, "d2") ~ "md2",
      str_detect(source, "d7") ~ "md7",
      TRUE ~ NA_character_
    )
  )


ols_plot$lhtime <- factor(ols_plot$lhtime, levels = c('0' ,'3' ,'5' ,'7' ,'10', '12', "na"))


ols_timeline <- ggplot(data=ols_plot%>% filter( lhtime != "na"), aes(x=lhtime, y=expression)) +
  
  geom_point(aes(fill=sigref,color = sigref, shape = shape), size= 6)+
  scale_fill_d3( breaks=c('h112', 'h120' ,'h124' , 'md2' , 'md7'), labels=c("Adult-H", "Infant-120", "Infant-124", "Day 2-M","Day 7-M")) +   scale_color_d3( breaks=c('h112', 'h120' ,'h124' , 'md2' , 'md7'), labels=c("Adult-H", "Infant-120", "Infant-124", "Day 2-M","Day 7-M"))+
  geom_boxplot(size=1.2, alpha = 0.35, fill='lightgrey') + 
  theme_classic(base_size = 30, base_family = "Myriad Pro") +
  theme(legend.position="bottom", axis.text.x = element_text(angle = 45, hjust = 1)) +
  xlab("Day in vitro")+
  ylab("Estimated Cell Proportion")+
  scale_y_continuous(breaks = c(0,0.25, 0.5, 0.75, 1),limits = c(-0.05,1))+
  scale_shape_manual(values = c("X" = 4, "point" = 23))+
  labs(color = "Single Cell Reference") +
  guides(color = guide_legend(title.position = "bottom"))+
  facet_wrap(~name, nrow=2, ncol = 2)+
  theme(panel.spacing = unit(1.2, "lines"))
ols_timeline

#ggsave(ols_timeline, dpi = 600, filename = "~/OneDrive - UW/data for deconv/Paper revision materials/updated_figures/testis_timeline_OLS_ALRA.png", width = 12, height=14, units = 'in', bg = 'white')

#ggsave(ols_timeline, dpi = 600, filename = "~/OneDrive - UW/data for deconv/Paper revision materials/updated_figures/testis_LEGEND.png", width = 30, height=14, bg = 'white')


```







# FOR LH PLOTS

```{r OLS alra LH}

ols_plot$time <- factor(ols_plot$time, levels = c('0','2','12','48'))

LH_ols_ALRA <- ggplot(data=ols_plot%>% filter(imp== 'ALRA' & exp =='lh' & name == "Leydig/Stroma cells"), aes(x=time, y=expression)) +
  
  geom_point(aes(fill=sigref,color = sigref, shape = shape), size= 6)+
  scale_fill_d3( breaks=c('h112', 'h120' ,'h124' , 'md2' , 'md7'), labels=c("Adult-H", "Infant-120", "Infant-124", "Day 2-M","Day 7-M")) + scale_color_d3( breaks=c('h112', 'h120' ,'h124' , 'md2' , 'md7'), labels=c("Adult-H", "Infant-120", "Infant-124", "Day 2-M","Day 7-M"))+
  geom_boxplot(size=1.2, alpha = 0.35, fill='lightgrey') + 
  theme_classic(base_size = 30, base_family = "Myriad Pro") +
  theme(legend.position="bottom", axis.text.x = element_text(angle = 45, hjust = 1)) +
  xlab("Time after LH Stimulus on Day 10 in vitro")+
  ylab("Estimated Cell Proportion")+
  scale_y_continuous(breaks = c(0,0.25, 0.5, 0.75, 1), limits = c(-0.05,1))+
  scale_shape_manual(values = c("X" = 4, "point" = 23))+
  scale_x_discrete(labels = c("0" = "Baseline",
                            "2" = "2 hours",
                            "12" = "12 hours",
                            "48" = "48 hours"))+
  labs(color = "Single Cell Reference") +
  guides(color = guide_legend(title.position = "bottom"))+
  facet_wrap(~name, nrow=1, ncol = 1)+
  theme(panel.spacing = unit(1.2, "lines"))
LH_ols_ALRA

#ggsave(LH_ols_ALRA, dpi = 600, filename = "~/OneDrive - UW/data for deconv/Paper revision materials/updated_figures/testis_LH_ols-alra.png", width = 10, height=10, units = 'in', bg = 'white')



```

```{r dwls alra LH}

dwls_testis$time <- factor(dwls_testis$time, levels = c('0','2','12','48'))

LH_DWLS_ALRA <- ggplot(data=dwls_testis%>% filter(imp== 'ALRA' & exp =='lh' & name == "Leydig/Stroma cells"), aes(x=time, y=expression)) +
  
  geom_point(aes(fill=sigref,color = sigref, shape = shape), size= 6)+
  scale_fill_d3( breaks=c('h112', 'h120' ,'h124' , 'md2' , 'md7'), labels=c("Adult-H", "Infant-120", "Infant-124", "Day 2-M","Day 7-M")) + scale_color_d3( breaks=c('h112', 'h120' ,'h124' , 'md2' , 'md7'), labels=c("Adult-H", "Infant-120", "Infant-124", "Day 2-M","Day 7-M"))+
  geom_boxplot(size=1.2, alpha = 0.35, fill='lightgrey') + 
  theme_classic(base_size = 30, base_family = "Myriad Pro") +
  theme(legend.position="bottom", axis.text.x = element_text(angle = 45, hjust = 1)) +
  xlab("Time after LH Stimulus on Day 10 in vitro")+
  ylab("Estimated Cell Proportion")+
  scale_y_continuous(breaks = c(0,0.25, 0.5, 0.75, 1), limits = c(-0.05,1))+
  scale_shape_manual(values = c("X" = 4, "point" = 23))+
  scale_x_discrete(labels = c("0" = "Baseline",
                            "2" = "2 hours",
                            "12" = "12 hours",
                            "48" = "48 hours"))+
  labs(color = "Single Cell Reference") +
  guides(color = guide_legend(title.position = "bottom"))+
  facet_wrap(~name, nrow=1, ncol = 1)+
  theme(panel.spacing = unit(1.2, "lines"))
LH_DWLS_ALRA

#ggsave(LH_DWLS_ALRA, dpi = 600, filename = "~/OneDrive - UW/data for deconv/Paper revision materials/updated_figures/testis_LH_dwls-alra.png", width = 10, height=10, units = 'in', bg = 'white')



```


```{r dwls noIMP LH}

summary(dwls_testis)

LH_DWLS_noimp <- ggplot(data=dwls_testis%>% filter(imp== 'noIMP' & exp =='lh' & name == "Leydig/Stroma cells"), aes(x=time, y=expression)) +
  
  geom_point(aes(fill=sigref,color = sigref, shape = shape), size= 6)+
  scale_fill_d3( breaks=c('h112', 'h120' ,'h124' , 'md2' , 'md7'), labels=c("Adult-H", "Infant-120", "Infant-124", "Day 2-M","Day 7-M")) + scale_color_d3( breaks=c('h112', 'h120' ,'h124' , 'md2' , 'md7'), labels=c("Adult-H", "Infant-120", "Infant-124", "Day 2-M","Day 7-M"))+
  geom_boxplot(size=1.2, alpha = 0.35, fill='lightgrey') + 
  theme_classic(base_size = 30, base_family = "Myriad Pro") +
  theme(legend.position="bottom", axis.text.x = element_text(angle = 45, hjust = 1)) +
  xlab("Time after LH Stimulus on Day 10 in vitro")+
  ylab("Estimated Cell Proportion")+
  scale_y_continuous(breaks = c(0,0.25, 0.5, 0.75, 1), limits = c(-0.05,1))+
  scale_shape_manual(values = c("X" = 4, "point" = 23))+
  scale_x_discrete(labels = c("0" = "Baseline",
                            "2" = "2 hours",
                            "12" = "12 hours",
                            "48" = "48 hours"))+
  labs(color = "Single Cell Reference") +
  guides(color = guide_legend(title.position = "bottom"))+
  facet_wrap(~name, nrow=1, ncol = 1)+
  theme(panel.spacing = unit(1.2, "lines"))
LH_DWLS_noimp

#ggsave(LH_DWLS_noimp, dpi = 600, filename = "~/OneDrive - UW/data for deconv/Paper revision materials/updated_figures/testis_LH_dwls-noIMP.png", width = 10, height=10, units = 'in', bg = 'white')


```



# Summary table

```{r summary}

testis_summary <- rbind(alra_dwls %>% mutate(method = 'DWLS-ALRA'), alra_OLS%>% mutate(method = 'OLS-ALRA'),noIMP_dwls%>% mutate(method = 'DWLS-noIMP'))%>% rownames_to_column("sample") %>% mutate(sample = str_replace_all(sample, "\\.\\..*$", ""))%>% pivot_longer(cols = c('PTM', 'Sertoli', 'germ_all', 'lc_all'), names_to = 'cell_type', values_to = 'expression') %>% mutate(tissue = if_else(grepl("827", sample), "tissue", "cells"))





testis_table <- testis_summary %>%
  group_by( method,cell_type) %>%
  summarise(
    mean = mean(expression),
    min = min(expression),
    max = max(expression),
    .groups = 'drop')%>%
    mutate(across(where(is.numeric), ~round(.x, 3)))



testis_tis_table <- testis_summary %>%
  group_by( tissue,method,cell_type) %>%
  summarise(
    mean = mean(expression),
    min = min(expression),
    max = max(expression),
    .groups = 'drop')%>%
    mutate(across(where(is.numeric), ~round(.x, 3)))

```

