
# Leveraging RNA-seq Deconvolution to Improve Complex In Vitro Model Characterization


### Brad C. Hansen, Christopher M. Arian, Yuting Zeng, Meg G. Takezawa, Ashleigh B. Theberge, Elaine M. Faustman, Kenneth E. Thummel, Edward J. Kelly

##### University of Washington, Seattle, Washington, U.S.A.


##### Brad Hansen and Christopher Arian contributed equally.
##### Edward Kelly is the corresponding author for this report.





#### Abstract:

Accelerating development of complex in vitro models (CIVMs) drives a need for additional methods to characterize these systems for use in toxicology and drug development. Relative to traditional cell culture, CIVMs can use a lower number of cells, are cultured for longer periods of time, and typically involve engineered 3-D cell organization. Standard single-cell assessment tools are not conducive to these types of complex models due to cell isolation stress, cell loss, and high cost. In this report, we benchmark RNA-seq deconvolution, which utilizes publicly available scRNA-seq datasets to predict cell proportions from bulk RNA-seq data derived from two CIVMs: a stem-cell-based human intestinal organoid CIVM and a neonatal rodent testis CIVM. We consider the impact of multiple imputation methods for scRNA-seq to restore the gene distribution of the original tissue and benchmark multiple deconvolution methods. The accuracy of deconvolution methods varied significantly in our analyses but provided valuable information on the emergence of an enterocyte cell population from the LGR5+ crypt stem cells following differentiation in the intestinal organoid CIVM. In the testis CIVM, deconvolution indicated that a small population of germ cells were retained over time, peritubular myoid cells proliferated over time, and that Leydig cell estimates remained stable with physiologically relevant hormone stimulation. In our analysis, using imputed single cell references improved deconvolution accuracy. Deconvolution can be a useful tool for novel CIVM characterization, especially with rapidly growing libraries of single-cell data across tissues and developmental time.

